src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs

```diff
@@ -267,48 +267,43 @@ private static void HandleLastWin32Error()
 
     internal static class NtProcessInfoHelper
     {
-        // Cache a single buffer for use in GetProcessInfos().
-        private static long[]? CachedBuffer;
-
         // Use a smaller buffer size on debug to ensure we hit the retry path.
 #if DEBUG
-        private const int DefaultCachedBufferSize = 1024;
+        private const int DefaultCachedBufferSize = 8 * 1024;
 #else
-        private const int DefaultCachedBufferSize = 128 * 1024;
+        private const int DefaultCachedBufferSize = 1024 * 1024;
 #endif
+        private static int MostRecentSize = DefaultCachedBufferSize;
 
         internal static ProcessInfo[] GetProcessInfos(int? processIdFilter = null)
         {
-            ProcessInfo[] processInfos;
-
             // Start with the default buffer size.
-            int bufferSize = DefaultCachedBufferSize;
-
-            // Get the cached buffer.
-            long[]? buffer = Interlocked.Exchange(ref CachedBuffer, null);
+            int bufferSize = MostRecentSize;
 
-            try
+            while (true)
             {
-                while (true)
-                {
-                    if (buffer == null)
-                    {
-                        // Allocate buffer of longs since some platforms require the buffer to be 64-bit aligned.
-                        buffer = new long[(bufferSize + 7) / 8];
-                    }
-
-                    uint requiredSize = 0;
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
 
+                try
+                {
                     unsafe
                     {
                         // Note that the buffer will contain pointers to itself and it needs to be pinned while it is being processed
                         // by GetProcessInfos below
-                        fixed (long* bufferPtr = buffer)
+                        fixed (byte* bufferPtr = buffer)
                         {
+                            // some platforms require the buffer to be 64-bit aligned.
+                            // ArrayPool does not guarantee 64-bit alignment, so we take the address of first 64-bit aligned element
+                            // we round up the address of the pinned array to the nearest multiple of 64
+                            Debug.Assert(bufferSize > 8);
+                            byte* alignedBufferPtr = (byte*)(((nint)bufferPtr + 7) & ~7);
+                            int firstAlignedElementIndex = (int)(alignedBufferPtr - bufferPtr);
```

Comment:
This is renting an ArrayPool array but then pinning it and trying to get an aligned pointer.  Should it just use NativeMemory.AllocAligned instead?

Reply:
Regular `NativeMemory.Alloc` would be enough. It guarantees sufficient alignment. This is trying to simulate alignment provided by `NativeMemory.Alloc` on Windows.

The strategy used by Windows heap for allocations larger than 512kB is pass through to VirtualAlloc/VirtualFree. It does not pool them in any way. In theory, using ArrayPool has speed performance advantage for repeated calls since it avoids page-faulting the pages in every time, at the cost of higher peak memory consumption. It would be useful to measure whether this advantage is material for the total cost of this method.

Reply:
Do you plan to measure this?